examples/postgres/custom_arrays/README.md

```diff
@@ -0,0 +1,892 @@
+        match *self {
+            ProtocolType::UnknownProtocol => out.write_all(b"UnknownProtocol")?,
+            ProtocolType::GRPC => out.write_all(b"GRPC")?,
+            ProtocolType::HTTP => out.write_all(b"HTTP")?,
+            ProtocolType::UDP => out.write_all(b"UDP")?,
+        }
+        Ok(IsNull::No)
+    }
+}
+
+impl FromSql<PgProtocolType, Pg> for ProtocolType {
+    fn from_sql(bytes: PgValue<'_>) -> deserialize::Result<Self> {
+        match bytes.as_bytes() {
+            b"UnknownProtocol" => Ok(ProtocolType::UnknownProtocol),
+            b"GRPC" => Ok(ProtocolType::GRPC),
+            b"HTTP" => Ok(ProtocolType::HTTP),
+            b"UDP" => Ok(ProtocolType::UDP),
+            _ => Ok(ProtocolType::UnknownProtocol),
+        }
+    }
+}
+```
+
+In toSql, it is important to end the implementation with Ok(IsNull::No). In the from_sql, it is important to add a catch
+all case that returns an UnknownProtocol instance. In practice, the catch all rarely ever gets triggered, but in those
+few corner cases when it does, it keeps the application running.
+
+### Rust Struct Endpoint
+
+```rust 
+#[derive(Debug, Clone, FromSqlRow, AsExpression, PartialEq, Eq)]
+#[diesel(sql_type=crate::schema::smdb::sql_types::ServiceEndpoint)]
+pub struct Endpoint {
+    pub name: String,
+    pub version: i32,
+    pub base_uri: String,
+    pub port: i32,
+    pub protocol: ProtocolType,
+}
+```
+
+It is worth mentioning is that the sql_type refers to the wrapper struct generated by Diesel and stored in the schema.rs
+file. Keep this in mind if you ever refactor the schema.rs file into a different folder because the macro annotation
+checks the path during compilation and throws an error if it cannot find the type in the provided path. The serialize
+and deserialize implementation of a custom type is not as obvious as the Enum because, internally, Postgres represent a
+custom type as an anonymous typed tuple. Therefore, you have to map a struct to a typed tuple and back.
+
+Let’s start with the toSql implementation to store the Rust Endpoint struct as typed tuple. Luckily, Diesel provides a
+helper util to do just that.
+
+```rust 
+impl ToSql<crate::schema::smdb::sql_types::ServiceEndpoint, Pg> for Endpoint {
+    fn to_sql<'b>(&'b self, out: &mut Output<'b, '_, Pg>) -> serialize::Result {
+        serialize::WriteTuple::<(Text, Integer, Text, Integer, PgProtocolType)>::write_tuple(
+            &(
+                self.name.to_owned(),
+                self.version.to_owned(),
+                self.base_uri.to_owned(),
+                self.port.to_owned(),
+                self.protocol.to_owned(),
+            ),
+            &mut out.reborrow(),
+        )
+    }
+}
+```
+
+I cannot stress enough that it is paramount that the tuple type signature must match exactly the Postgres type signature defined in your up.sql. Ideally, you want to use the split view function in your IDE to have the up.sql 
```

Comment:
Note for diesel-team/myself: We might want to look into generating that as part of the `schema.rs` file as well. Maybe just as `impl RecordLike { type Fields = (…)}` for record types and maybe for enums `impl EnumLike { const FIELDS: &str = &[…] }` 

(That would likely also result in diesel-cli generating the relevant SQL type for the enum here as well)

To be clear: That's nothing that you need to address as this PR, that's just me thinking loud.